src/clients/c++/perf_client/load_manager.cc

```diff
@@ -301,98 +544,167 @@ LoadManager::PrepareInfer(
-                  std::to_string(batch1_size) +
-                  " bytes for each batch, but provided data has " +
-                  std::to_string(it->second.size()) + " bytes");
-        }
-        data = (const uint8_t*)&(it->second)[0];
-      } else if (input_buf_.size() != 0) {
-        if (batch1_size > input_buf_.size()) {
-          return nic::Error(
-              ni::RequestStatusCode::INTERNAL,
-              "input '" + input->Name() + "' requires " +
-                  std::to_string(batch1_size) +
-                  " bytes for each batch, but generated data has " +
-                  std::to_string(input_buf_.size()) + " bytes");
-        } else {
-          data = &input_buf_[0];
-        }
-      } else {
-        return nic::Error(
-            ni::RequestStatusCode::INVALID_ARG,
-            "unable to find data for input '" + input->Name() + "'.");
-      }
-
       for (size_t i = 0; i < batch_size_; ++i) {
-        RETURN_IF_ERROR(input->SetRaw(data, batch1_size));
+        RETURN_IF_ERROR(input->SetRaw(data_ptr, batch1_size));
       }
     } else {
-      size_t batch1_num_strings = 1;
-      if (!input->Shape().empty()) {
-        for (const auto dim : input->Shape()) {
-          batch1_num_strings *= dim;
-        }
-      } else {
-        for (const auto dim : input->Dims()) {
-          batch1_num_strings *= dim;
-        }
-      }
-
-      bool used_new_allocation = false;
-      std::vector<std::string>* data;
-
-      // if available, use provided data instead
-      auto it = input_string_data_.find(input->Name());
-      if (it != input_string_data_.end()) {
-        if (it->second.size() != batch1_num_strings) {
-          return nic::Error(
-              ni::RequestStatusCode::INVALID_ARG,
-              "input '" + input->Name() + "' requires " +
-                  std::to_string(batch1_num_strings) +
-                  " strings for each batch, but provided data has " +
-                  std::to_string(it->second.size()) + " strings.");
-        }
-        data = &it->second;
-      } else if (input_string_buf_.size() != 0) {
-        if (batch1_num_strings > input_string_buf_.size()) {
-          return nic::Error(
-              ni::RequestStatusCode::INTERNAL,
-              "input '" + input->Name() + "' requires " +
-                  std::to_string(batch1_num_strings) +
-                  " strings for each batch, but generated data has " +
-                  std::to_string(input_string_buf_.size()) + " strings.");
-        } else {
-          used_new_allocation = true;
-          data = new std::vector<std::string>(
-              input_string_buf_.begin(),
-              input_string_buf_.begin() + batch1_num_strings);
-        }
-      } else {
-        return nic::Error(
-            ni::RequestStatusCode::INVALID_ARG,
-            "unable to find data for input '" + input->Name() + "'.");
-      }
+      std::shared_ptr<std::vector<std::string>> data;
```

Comment:
You can just make this a local variable instead of a smart pointer, I don't think you need to `new` the vector in `GetInputData()`.

Reply:
This was needed because the SetFromString() function reads data from a string vector. So for the inputs of different size I had to prepare the tensor. It would have caused a considerable slow down while opening contexts, because the API serializes the string input data. Hence, to avoid the extraneous string copies, I am now serializing the string data in initialization phase itself. Also, I am maintaining different copies for different string inputs in the model to facilitate easy retrieval of their serialized byte sizes.